### PR TITLE
expression: Do not wrap a 'CAST As JSON' when input is already a JSON

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: expression/builtin_cast_test.go [hunk 1]
```diff
@@ -25,11 +25,9 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/charset"
-	"github.com/pingcap/tidb/util/testleak"
 )
 
 func (s *testEvaluatorSuite) TestCast(c *C) {
-	defer testleak.AfterTest(c)()
```
Comment (winoros) [inline expression/builtin_cast_test.go]: Why remove this?

**Comment by shenli** [comment]
Please fix the CI.

**Comment by zz-jason** [comment]
@XuHuaiyu @lamxTyler PTAL

File: mysql/type.go [hunk 2]
```diff
@@ -52,28 +52,28 @@ const TypeUnspecified = TypeDecimal
 
 // Flag information.
 const (
```
Comment (XuHuaiyu) [inline mysql/type.go]: const (
     NotNullFlag uint = 1 << iota
     PriKeyFlag
     ....
)

Comment (zz-jason) [reply mysql/type.go]: `NumFlag` and `GroupFlag` have the same value, I think maybe we should hand code the values of these flags.

Comment (XuHuaiyu) [reply mysql/type.go]: ok
mysql src code reference: https://github.com/mysql/mysql-server/blob/8.0/include/mysql_com.h#L164-L166

**Pull request merged**